lib/ocpp/v201/charge_point.cpp

```diff
@@ -563,6 +589,113 @@ AuthorizeResponse ChargePoint::validate_token(const IdToken id_token, const std:
+                        response.idTokenInfo.status = AuthorizationStatusEnum::Invalid;
+                    }
+                } else {
+                    // Try to generate the OCSP data from the certificate chain and use that
+                    std::vector<OCSPRequestData> generated_ocsp_request_data_list =
+                        generate_ocsp_data(certificate.value());
+                    if (generated_ocsp_request_data_list.size() > 0) {
+                        EVLOG_info << "Online: Pass generated OCSP data to CSMS";
+                        response = this->authorize_req(id_token, std::nullopt, generated_ocsp_request_data_list);
+                        forwardToCsms = true;
+                    } else {
+                        EVLOG_warning << "Online: OCSP data could not be generated";
+                        response.idTokenInfo.status = AuthorizationStatusEnum::Invalid;
+                    }
+                }
+            } else { // Offline
+                // C07.FR.08: CS shall try to validate the contract locally
+                if (ContractValidationOffline) {
+                    EVLOG_info << "Offline: contract " << localVerifyResult;
+                    switch (localVerifyResult) {
+                    // C07.FR.09: CS shall lookup the eMAID in Local Auth List or Auth Cache when
+                    // local validation succeeded
+                    case CertificateValidationResult::Accepted:
+                        // In C07.FR.09 LocalAuthorizeOffline is mentioned, this seems to be a generic config item
+                        // that applies to Local Auth List and Auth Cache, but since there are no requirements about
+                        // it, lets check it here
+                        if (LocalAuthorizeOffline) {
+                            tryLocalAuthListOrCache = true;
+                        } else {
+                            // No requirement states what to do when ContractValidationOffline is true
+                            // and LocalAuthorizeOffline is false
+                            response.idTokenInfo.status = AuthorizationStatusEnum::NotAtThisTime;
+                            response.certificateStatus = AuthorizeCertificateStatusEnum::Accepted;
+                        }
+                        break;
+                    case CertificateValidationResult::Expired:
+                        response.idTokenInfo.status = AuthorizationStatusEnum::Expired;
+                        response.certificateStatus = AuthorizeCertificateStatusEnum::CertificateExpired;
+                        break;
+                    default:
+                        response.idTokenInfo.status = AuthorizationStatusEnum::Invalid;
+                        break;
+                    }
+                } else {
+                    // C07.FR.07: CS shall not allow charging
+                    response.idTokenInfo.status = AuthorizationStatusEnum::NotAtThisTime;
```

Comment:
I would suggest a different return type here. `NotAtThisTime` holds a special meaning an could be used for UI purposes later. In this case I would just go with `Unknown`.

Reply:
I have set it to `NotAtThisTime` for this purpose as the CSMS is offline. If the user comes back some other time if the CSMS if online again, authorization might be ok. The user gets feedback that the failed authorization is because of some problem in the CS, not with his contract.

Reply:
Fair enough but that would be for other modules to decide what to show to the used based on the online status. NotAtThisTime is a specific return type specified in OCPP and in my opinion should hold special meaning if the CSMS reports this.

In other cases when we are offline and not able to authorize we also respond with unknown.

Reply:
ok, done